### DISCUSSION TITLE
Can't seem to extend themes anymore - alternative to extendTheme?

### DISCUSSION THREAD

**Discussion by mjemerson** [chakra-ui/chakra-ui#9016] [Q&A]
Hello,

I've been trying to migrate from 2.8 to 3.0 and having a lot of difficulties with theming. Currently I have several separate files containing various variants for components, such as buttons, accordions, alerts etc. Before I used this structure to implement them into my project:

```
const ChakraThemeProvider: FC<PropsWithChildren> = ({ children }) => {
  const tokens = useTokens();

  const theme = extendTheme({
    components: {
      Alert: alertStyles(tokens),
      Accordion: accordionStyles(tokens)
      ....
```

However, this method seems deprecated now and there only appears to be `createSystem` which does not seem to work as expected. According to the documentation, I'd have to stipulate all my variants in one file (which seems completely ineffective to me) unless I'm misunderstanding the way it's supposed to work?

Any help with this would be greatly appreciated as currently trying to migrate my rather large project is becoming more and more difficult.

Thank you
Michael

**Comment by segunadebayo** [comment]
Could you share a reproduction of v2 and I can help with the v3 version.

**Reply by mjemerson** [reply]
Well, my post is pretty much it. Each of those files (alertStyles, accordionStyles) is a set of styling variants for each component, but they use a private library of tokens that define colour, sizes etc.

Example:

```
import type { Tokens } from '../context/TokenProvider';
import { makeMed3, makeReg4 } from './textStyles';

const alertStyles = (tokens: Tokens) => ({
  variants: {
    success: {
      container: {
        padding: tokens.Spacing05,
        backgroundColor: tokens.SuccessLight,
        borderColor: tokens.BorderSuccess,
        borderBottomWidth: tokens.BorderWidth02,
      },
      description: makeReg4(tokens, {
        color: tokens.TextDark800,
      }),
      icon: {
        color: tokens.TextSuccess,
        alignSelf: 'start',
      },
      title: makeMed3(tokens, {
        color: tokens.TextDark800,
      }),
    },
}
```

After all these styles are imported (there's about 20 of them) the ChakraProvider is return as such:

`return <ChakraProvider theme={theme}>{children}</ChakraProvider>;`

Where `theme` is the `extendTheme` method encapsulating all the imports, as my original post.

**Reply by zgid123** [reply]
in v2, we can use a method and can get props from component to combine styling based on them, can we archive this one for v3?

```ts
export const Preloader = definePreloaderStyleConfig({
  baseStyle: definePreloaderPartsStyle(({ fullWidth, center }) => {
    let wrapperStyle: SystemStyleObject = {};

    if (fullWidth) {
      wrapperStyle = {
      };
    }

    if (center) {
      mergeStyle(wrapperStyle, {
      });
    }

    return {
      wrapper: wrapperStyle,
    };
  }),
});
```

**Reply by Daidalos117** [reply]
@zgid123 I am also curios how to do this, did you figure it out?

**Reply by mjemerson** [reply]
> @zgid123 I am also curios how to do this, did you figure it out?

Nope. Ended up downgrading back to 2.8, I don't think there's enough documentation / examples for 3.0 just yet. Not worth implementing a version without suitable documentation

**Reply by zgid123** [reply]
> @zgid123 I am also curios how to do this, did you figure it out?

I don't think we can do it at the moment. Maybe wait for the update from Chakra UI, I did rollback to version 2 already

**Reply by Daidalos117** [reply]
oh no 🫤 I must say the v3 documentation really missing some crucial things. But I already spend hours with upgrade, so going back is not an option. I probably will need to just pause it. I must say I am overally highly disappointed with v3 so far

**Reply by mjemerson** [reply]
> oh no 🫤 I must say the v3 documentation really missing some crucial things. But I already spend hours with upgrade, so going back is not an option. I probably will need to just pause it. I must say I am overally highly disappointed with v3 so far

Luckily in all our projects we have various repos on GitHub so we just reverted to an earlier version of the code and ran `npm install` with Chakra 2 in the package

**Reply by Daidalos117** [reply]
@segunadebayo when we can expect to be these things documented? I am missing so much info from v2

**Reply by rossanodr** [reply]
I also had trouble trying to migrate. I have no idea how to make the changes in the theme file, like handling custom Google fonts, for example.
`const theme = extendTheme({
    fonts: {
        roboto: roboto.style.fontFamily,
        lora: lora.style.fontFamily,
        poppins: poppins.style.fontFamily,
    },
    colors: customColors,
    semanticTokens: {
        colors: customColors,
    },

    components: {
        Textarea: textareaTheme,
        Accordion: accordionTheme,
        Form: formFloatingLabel,
        Tooltip: {
            baseStyle: {
                rounded: "lg",
            },
        },
    },

    config: {
        cssVarPrefix: "ck",
    },
}) as Theme;
`
This is my current file, and I have no idea how to update it for version 3.0. I ended up downgrading to version 2.8, and I’ll wait for the new documentation.

**Reply by filipeciandt** [reply]
I could create my custom theme like this on version 3.0.

### theme.ts
`````
import { createSystem, defaultConfig, defineConfig, defineAnimationStyles } from "@chakra-ui/react"

const customConfig = defineConfig({
  theme: {
    recipes: {
      text: {
        base: {
          color: 'GrayText'
        }
      },
      button: {
        base: {
          // your custom styles
        }
      },
    },
    tokens: {
      colors: {
        myProjectName: {
          myColorCustomName: {
            DEFAULT: { value: '#FA5A50' },
            100: { value: '#FAB8B4' },
            200: { value: '#FA8982' },
            300: { value: '#FA5A50' },
            400: { value: '#D94A41' },
            500: { value: '#A63832' },
          },
        }
      },
      fontWeights: {
        DEFAULT: { value: '200' },
        normal: { value: "200" },
        semibold: { value: "500" },
        bold: { value: "800" },
      },
      fonts: {
        body: { value: 'Plus Jakarta Sans' }
      },
    },
    animationStyles: defineAnimationStyles({
      slideFadeIn: {
        value: {
          animationName: "slide-from-right, fade-in",
          animationDuration: "0.2s",
          animationTimingFunction: "ease-in-out",
          animationIterationCount: "1",
        },
      },
      slideFadeOut: {
        value: {
          animationName: "slide-to-right, fade-out",
          animationDuration: "0.2s",
          animationTimingFunction: "ease-in-out",
          animationIterationCount: "1",
        },
      },
    }),
  },
  globalCss: {
    "html, body": {
      margin: 0,
      padding: 0,
    },
    body: {
      background: "#FCFCFD",
      overflow: 'hidden'
    },
    '::-webkit-scrollbar': {
      bg: 'transparent',
      h: '8px',
      w: '8px'
    },
    '::-webkit-scrollbar-thumb': {
      border: 'none',
      boxShadow: 'none',
      background: 'gray.200',
      borderRadius: '4px',
      minHeight: '40px',
    }
  },
})

export const customSystem = createSystem(defaultConfig, customConfig)
```

**Reply by filipeciandt** [reply]
If you see chakra engine does not react to your changes maybe you need to run the chakra typegen to generate types.

`npx @chakra-ui/cli typegen path/to/your/theme.ts`

**Reply by zgid123** [reply]
@Daidalos117 from the core team

> it's recommended to use selectors / conditions for that, eg with data- attributes. But themes also have more flexibility now because many cases you can solve now with custom variants

It seems we only have two options to archive what we did.

1. we define a list of conditions, then instead of use multiple props to combine the style, we will convert those things into a list states. [Ref](https://www.chakra-ui.com/docs/theming/customization/conditions)
2. instead of using conditions, we will use custom variant or compound variants to archive that

We cannot use prop like we did before. I don't like this approach, that's why I asked the question on discord, but still this is the new approach from the team, we need to change the approach to archive what we want

**Reply by maulik13** [reply]
> oh no 🫤 I must say the v3 documentation really missing some crucial things. But I already spend hours with upgrade, so going back is not an option. I probably will need to just pause it. I must say I am overally highly disappointed with v3 so far

Documentation in general is hard for a project like this. But the team has documented really well. What it is missing is migration for some of the old features, which I imagine is not easy to do.  Once you manage to move over I am sure you will change your mind about v3 :)

**Reply by surielmx** [reply]
In my case I'm not migrating a project with v2, I'm creating it from scratch with v3, but it's a pain. A lot of configurations to make things work. Years ago things weren't so complicated and they were easier to configure to have a project ready. I've been trying all day to get the project running with v3, but I find errors or things that don't work as the documentation says. I think I'll keep using v2.

**Reply by maulik13** [reply]
It is not fun to have a project not work out of the box. But, it worked for me. I managed to get my migration from v2 to v3 done in 2 days and I removed a lot of customization since the default worked great out of the box. I guess you will just have to ask questions and help the team get documentation fixed if something is not consistent.

**Comment by heikkilae** [comment]
I'm also looking for example to extend Button with new variants without loosing the default variants. So far I have found only the following textStyles example from [documentation](https://next.chakra-ui.com/docs/styling/text-styles#update-the-theme):
```
import { createSystem, defineConfig } from "@chakra-ui/react"
import { textStyles } from "./text-styles"

const config = defineConfig({
  theme: {
    extend: {
      textStyles,
    },
  },
})

export default createSystem(defaultConfig, config)
```

However it seems not to work for me and I'm also having type script error:
```
Object literal may only specify known properties, and 'extend' does not exist in type 'ThemingConfig'.ts(2353)
types.d.ts(230, 5): The expected type comes from property 'theme' which is declared here on type 'SystemConfig'
(property) extend: {
    textStyles: TextStyles;
}
```

I'm using  `"@chakra-ui/react": "^3.1.0" `currently.

**Reply by heikkilae** [reply]
I guess I found the V3 way to customize default theme with [mergeConfigs](https://next.chakra-ui.com/docs/theming/customization/overview#customization):
```
export const buttonRecipe = defineRecipe({
  variants: {
    variant: {
      // Override variant from default recipe
      solid: {
        bg: "blue.500",
        color: "white",
      },

      // Add a new variant
      custom: {
        bg: "blue.800",
        color: "white",
      },
    },
  },
});

const customConfig = defineConfig({
  theme: {
    recipes: {
      button: buttonRecipe,
    },
  },
});

const config = mergeConfigs(defaultConfig, customConfig);

export default createSystem(config);
```


```
import system from "@/theme";

<ChakraProvider value={system}>
  {children}
</ChakraProvider>
```

**Reply by Daidalos117** [reply]
@heikkilae that's just the same as 
```export default createSystem(defaultConfig, config)```  ?
I think the createSystem use mergeConfigs internally

**Reply by heikkilae** [reply]
The issue with `export default createSystem(defaultConfig, config)` is that it doesn't generate types correctly and button doesn't accept `variant="custom"` even though I run typegen. With mergeConfigs and typegen the new variant works after TS server restart.

**Reply by Daidalos117** [reply]
@heikkilae  actually you are right, that seems to work! I was looking for that when I started to migrate 😅

**Reply by ericksonholguin** [reply]
@Daidalos117 @heikkilae I'm trying do that, but I can't get it, Did you do another step?

**Reply by zgid123** [reply]
> @Daidalos117 @heikkilae I'm trying do that, but I can't get it, Did you do another step?

you mean run the script typegen? Remember `export default createSystem(theme)`

**Reply by ericksonholguin** [reply]
> > @Daidalos117 @heikkilae I'm trying do that, but I can't get it, Did you do another step?
> 
> you mean run the script typegen? Remember `export default createSystem(theme)`

@zgid123 Yes, I opened a issue for this problem https://github.com/chakra-ui/chakra-ui/issues/9168, you can say there all details

**Reply by Sanch0ous** [reply]
Been stuck on this for hours, and after using mergeConfigs and run typegen update, it works like a charm !

**Reply by heikkilae** [reply]
> The issue with `export default createSystem(defaultConfig, config)` is that it doesn't generate types correctly and button doesn't accept `variant="custom"` even though I run typegen. With mergeConfigs and typegen the new variant works after TS server restart.

@segunadebayo: Can you confirm that my [mergeConfigs approach](https://github.com/chakra-ui/chakra-ui/discussions/9016#discussioncomment-11174648) is the correct way extend default theme? Wondering if `createSystem(defaultConfig, customConfig)` should be enough but it somehow doesn't work for me.

**Reply by zgid123** [reply]
@heikkilae yup, the code is correct. I just come back to migrate my project to V3, and just figure it out that we need to pass the defaultConfig to merge our custom config to their config or else we cannot use the shorthand properties. 

`createSystem` will use `mergeConfigs` internally
```ts
export function createSystem(...configs: SystemConfig[]): SystemContext {
  const config = mergeConfigs(...configs)
  // other code
}
``` 

But it seems at the moment, we cannot use `_dark` prop like other prop unless we use nested one

```ts
{
  _dark: { // will raise error for typescript but still work
    [customColorVar]: 'currentColor',
  },
  bgColor: {
    base: 'red.400',
    _dark: 'red.600' // won't raise error 
  }
}
```

I think the typing is incorrect at the moment

**Comment by bolebon** [comment]
Hi guys, I actually ran into  this Github issue when asking for help on Discord.
Here's an example of how you can extend / override the style of the components in Chakra UI V3.

```ts
const inputRecipe = defineRecipe({
  variants: {
    variant: {
      outline: {
        bg: "bg",
      },
    },
  },
})

const selectRecipe = defineSlotRecipe({
  slots: ["trigger"],
  variants: {
    variant: {
      outline: {
        trigger: {
          bg: "bg"
        }
      }
    }
  }
});

const tableRecipe = defineSlotRecipe({
  slots: ["root"],
  variants: {
    variant: {
      outline: {
        root: {
          bg: "bg",
          rounded: "sm",
        }
      }
    }
  }
});

export default createSystem(defaultConfig, {
  theme: {
    recipes: {
      input: inputRecipe,
    },
    slotRecipes: {
      select: selectRecipe,
      table: tableRecipe,
    },
  },
});
```
This uses a new mechanism called "recipe" that is explained in the documentation [here](https://www.chakra-ui.com/docs/theming/recipes) and [here](https://www.chakra-ui.com/docs/theming/slot-recipes).
There are "regular" recipes for single components and slot recipes for components composed of multiple sub components.

To better understand how the recipe of a given component is structured you have 2 possibilities:
- On the documentation of the component itself, you have a "Recipe" link at the top of the page which will take you to the code of the recipe.
- You can use this command: `npx @chakra-ui/cli eject --outdir theme` which will generate the code of the default theme which will help you understand how it is structured and, thus, how it can be overriden.

In any case, when overriding the styles of a component, all you have to do is replicate the structure of the default recipe and only define the styles that you want to add or override (they'll get merged with the default styles).

Hope this helps !

**Reply by heikkilae** [reply]
I can confirm this now. Have they fixed it somehow? At the time I wrote https://github.com/chakra-ui/chakra-ui/discussions/9016#discussioncomment-11221662, it wasn't working without mergeConfigs. And based on other replies, I believe it wasn't just me.

**Reply by Oddward** [reply]
I don't get what this is pointing to, the docs make it seem like I only need to add the definition for the variants I want, while that seems to exclude the existing variants when generating types. With your tip I tried copying the entire recipe from the repo and editing that instead but it only partially works for me, it adds the types but the styles are completely ignored. Colours and other stuff in the config work only the recipes do nothing.

E.g. 
1. copy the cardSlotRecipe from the repo linked at the top of the [card doc](https://www.chakra-ui.com/docs/components/card)
2. add variant under variant key with some bg
3. change bg value in outline variant
4. add cardSlotRecipe into customConfig
5. `export default createSystem(defaultConfig, customConfig)`
6. apply outline variant and it still shows the default bg, apply my variant and there's no bg

**Reply by zgid123** [reply]
@Oddward do you have a repo or code example?

**Reply by otahirs** [reply]
what worked for me was

1. 
```
const config = defineConfig({
  theme: {
    recipes: {
      button: buttonRecipe,
    },
  },
});

const system = createSystem(mergeConfigs(defaultConfig, config));
export default system;
```
2. run run the `chakra typegen` https://www.chakra-ui.com/docs/get-started/cli#chakra-typegen
3. restart TS server
3. profit

**Comment by GradyTruitt** [comment]
To see a full, working example of how to customize your system and add new variants to recipes (like button) you can run the `chakra eject --outdir src/theme` command from Chakra CLI. This generates all the code for the default system. You can modify as you'd like. You can replace the full recipe as this does, or you can just use this as a reference for how to accomplish customizations for certain components/recipes.

Note that when you change a recipe to add or modify the variants you must re run the typegen command to generate new types that include your variant changes `chakra typegen src/theme.ts` replacing `src/theme.ts` with the path to your theme root.

For me, I used this as a starting point, and built a fully custom theme for all of the chakra components. It was very helpful to see how something should be modified or changed.
